addons/mail/models/mail_thread.py

```diff
@@ -2538,40 +2624,69 @@ def _notify_classify_recipients(self, recipient_data, model_name, msg_vals=None)
 
         # fill group_data with default_values if they are not complete
         for group_name, group_func, group_data in groups:
-            group_data.setdefault('notification_group_name', group_name)
-            group_data.setdefault('notification_is_customer', False)
+            group_data.setdefault('active', True)
+            group_data.setdefault('actions', list())
             group_data.setdefault('has_button_access', True)
+            group_data.setdefault('notification_is_customer', False)
+            group_data.setdefault('notification_group_name', group_name)
+            group_data.setdefault('recipients', list())
             group_button_access = group_data.setdefault('button_access', {})
             group_button_access.setdefault('url', access_link)
             group_button_access.setdefault('title', view_title)
-            group_data.setdefault('actions', list())
-            group_data.setdefault('recipients', list())
 
         # classify recipients in each group
         for recipient in recipient_data:
             for group_name, group_func, group_data in groups:
-                if group_func(recipient):
+                if group_data['active'] and group_func(recipient):
                     group_data['recipients'].append(recipient['id'])
                     break
 
-        result = []
-        for group_name, group_method, group_data in groups:
-            if group_data['recipients']:
-                result.append(group_data)
+        # filter out groups without recipients
+        return [group_data for _group_name, _group_func, group_data in groups
+                if group_data['recipients']]
 
-        return result
+    @api.model
+    def _notify_encode_link(self, base_link, params):
+        secret = self.env['ir.config_parameter'].sudo().get_param('database.secret')
+        token = '%s?%s' % (base_link, ' '.join('%s=%s' % (key, params[key]) for key in sorted(params)))
+        hm = hmac.new(secret.encode('utf-8'), token.encode('utf-8'), hashlib.sha1).hexdigest()
```

Comment:
I guess we don't want break existing tokens? In case we did not cared, could be an oportunity to use the helper in tools but probably not worth it here